### PR TITLE
Member roles

### PR DESCRIPTION
**Pull request by grolu**
**What this PR does / why we need it**:
This PR adds the possibility to edit member roles in the UI

Screen Shot 2020-03-30 at 09 43 18

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```noteworthy user
Project member roles can now be configured in the member section
```

### CODE REVIEW

**Comment by petersutter** [comment]
call selfsubjectrulesreview after changing the roles of own user

File: backend/lib/routes/members.js [hunk 1]
```diff
@@ -55,6 +55,17 @@ router.route('/:name')
       next(err)
     }
   })
+  .put(async (req, res, next) => {
+    try {
+      const user = req.user
+      const namespace = req.params.namespace
+      const name = req.params.name
+      const roles = req.body.roles
+      res.send(await members.update({ user, namespace, name, roles }))
```
Comment (holgerkoser) [inline backend/lib/routes/members.js]:
```suggestion
      const body = req.body
      res.send(await members.update({ user, namespace, name, body }))
```
The route should not know about the body structure

File: backend/lib/services/members.js [hunk 2]
```diff
@@ -44,9 +44,9 @@ function fromResource (project = {}, serviceAccounts = []) {
     .chain(project)
     .get('spec.members')
     .filter(['kind', 'User'])
-    .map('name')
-    .map(username => ({
+    .map(({ name: username, role, roles }) => ({
```
Comment (holgerkoser) [inline backend/lib/services/members.js]: Don't understand that role, roles thing. Is this done beacuse of backward compatibility?

File: backend/lib/services/members.js [hunk 3]
```diff
@@ -44,9 +44,9 @@ function fromResource (project = {}, serviceAccounts = []) {
...
       username,
+      roles: roles ? [role, ...roles] : [role],
```
Comment (holgerkoser) [inline backend/lib/services/members.js]: What if role is empty?

File: backend/lib/services/members.js [hunk 4]
```diff
@@ -76,19 +76,21 @@ async function deleteServiceaccount (client, { namespace, name }) {
...
-async function setProjectMember (client, { namespace, name }) {
+async function setProjectMember (client, { namespace, name, roles }) {
   // get project
   const project = await client.getProjectByNamespace(namespace)
   // get project members from project
   const members = _.slice(project.spec.members, 0)
   if (_.find(members, ['name', name])) {
     throw new Conflict(`User '${name}' is already member of this project`)
   }
+  const role = roles.shift()
```
Comment (holgerkoser) [inline backend/lib/services/members.js]: what is special about the first role?

File: backend/lib/services/members.js [hunk 5]
```diff
@@ -98,6 +100,26 @@ async function setProjectMember (client, { namespace, name }) {
   return client['core.gardener.cloud'].projects.mergePatch(project.metadata.name, body)
 }
 
+async function updateProjectMember (client, { namespace, name, roles }) {
```
Comment (holgerkoser) [inline backend/lib/services/members.js]: `updateProjectMemberRoles`

File: backend/lib/services/members.js [hunk 6]
```diff
@@ -152,6 +174,7 @@ exports.get = async function ({ user, namespace, name }) {
     const clusterName = 'garden'
     const contextName = `${clusterName}-${projectName}-${name}`
     member.kind = 'ServiceAccount'
+    member.roles = member.roles ? [member.role, ...member.roles] : [member.role]
```
Comment (holgerkoser) [inline backend/lib/services/members.js]: What is with `member.role`? Why don't we crerate a new object?

File: backend/lib/services/members.js (hunk 2 above)
Comment (petersutter) [reply backend/lib/services/members.js]: yes

File: backend/lib/services/members.js [hunk 7]
```diff
@@ -165,7 +188,7 @@ exports.get = async function ({ user, namespace, name }) {
   return member
 }
 
-exports.create = async function ({ user, namespace, body: { name } }) {
+exports.create = async function ({ user, namespace, body: { name, roles } }) {
```
Comment (holgerkoser) [inline backend/lib/services/members.js]: That's is good. Do it in this way also in the update case.

File: backend/test/acceptance/api.members.spec.js [hunk 8]
```diff
@@ -75,28 +75,44 @@ module.exports = function ({ agent, k8s, auth }) {
...
+    k8s.stub.updateMember({ bearer, namespace, name, roles })
+    const res = await agent
+      .put(`/api/namespaces/${namespace}/members/${name}`)
+      .set('cookie', await user.cookie)
+      .send({ metadata, roles })
+    expect(res).to.have.status(200)
+    expect(res).to.be.json
+    const member = _.find(res.body, { username: 'bar@example.org' })
+    expect(member).to.eql({ username: 'bar@example.org', roles: ['newRole'] })
+  })
```
Comment (holgerkoser) [inline backend/test/acceptance/api.members.spec.js]:
```suggestion
   it('should update roles of a project member', async function () {
    const bearer = await user.bearer
    const name = 'bar@example.org'
    const roles = ['newRole']
    k8s.stub.updateMember({ bearer, namespace, name, roles })
    const res = await agent
      .put(`/api/namespaces/${namespace}/members/${name}`)
      .set('cookie', await user.cookie)
      .send({ metadata, roles })

    expect(res).to.have.status(200)
    expect(res).to.be.json
    const member = _.find(res.body, ['username', name])
    expect(member).to.eql({ username: name, roles })
  })
```

File: backend/test/acceptance/api.members.spec.js [hunk 9]
```diff
@@ -75,28 +75,44 @@ module.exports = function ({ agent, k8s, auth }) {
   it('should add a project member', async function () {
     const bearer = await user.bearer
     const name = 'baz@example.org'
-    k8s.stub.addMember({ bearer, namespace, name })
+    const roles = ['admin', 'owner']
```
Comment (holgerkoser) [inline backend/test/acceptance/api.members.spec.js]: same as below. Use defined constant values to compare.

File: backend/test/support/nocks/k8s.js [hunk 10]
```diff
@@ -291,7 +307,7 @@ function getProjectMembers (project) {
   return _
     .chain(project)
     .get('spec.members')
-    .map(({ name: username }) => ({ username }))
+    .map(({ name: username, role, roles }) => ({ username, roles: roles ? [role, ...roles] : [role] }))
```
Comment (holgerkoser) [inline backend/test/support/nocks/k8s.js]: Very hard to understand what is happing here?

File: backend/test/support/nocks/k8s.js [hunk 11]
```diff
@@ -1387,16 +1433,16 @@ const stub = {
...
+            verbs: ['get'],
+            apiGroups: ['core.gardener.cloud'],
+            resources: ['projects'],
+            resourceName: ['foo']
+          },
+          {
+            verbs: ['create'],
+            apiGroups: ['core.gardener.cloud'],
+            resources: ['projects']
+          }
```
Comment (holgerkoser) [inline backend/test/support/nocks/k8s.js]: Removed correct indentation!

File: backend/test/support/nocks/k8s.js [hunk 12]
```diff
@@ -1405,7 +1451,7 @@ const stub = {
             resources: ['projects'],
             resourceName: ['foo']
           }
-        ])
+          ])
```
Comment (holgerkoser) [inline backend/test/support/nocks/k8s.js]: indentation not correct.

File: frontend/src/components/ProjectServiceAccountRow.vue [hunk 13]
```diff
@@ -51,6 +52,13 @@ limitations under the License.
           {{username}}
         </v-list-tile-sub-title>
       </v-list-tile-content>
+      <v-list-tile-action>
+        <v-layout row align-center>
+          <v-chip v-for="(roleName, index) in roleNames" :key="index" small color="blue-grey darken-2" text-color="white">
```
Comment (holgerkoser) [inline frontend/src/components/ProjectServiceAccountRow.vue]: in the backend we talk about role and roles. What is the difference between `role` and `roleName`? Use `roleName` as key.

File: frontend/src/components/ProjectServiceAccountRow.vue [hunk 14]
```diff
@@ -134,18 +154,21 @@ export default {
...
     },
     createdByClasses () {
-      return !!this.createdBy ? ['font-weight-bold'] : ['grey--text']
+      return this.createdBy ? ['font-weight-bold'] : ['grey--text']
     }
   },
   methods: {
     async onDownload () {
-      this.$emit('onDownload', this.username)
+      this.$emit('download', this.username)
```
Comment (holgerkoser) [inline frontend/src/components/ProjectServiceAccountRow.vue]: async makes no sense

File: frontend/src/components/ProjectUserRow.vue [hunk 15]
```diff
@@ -32,6 +32,21 @@ limitations under the License.
           <span v-else class="pl-2">{{username}}</span>
         </v-list-tile-sub-title>
       </v-list-tile-content>
+      <v-list-tile-action>
+        <v-layout row align-center>
+          <v-chip v-for="(roleName, index) in roleNames" :key="index" small color="green darken-2" text-color="white">
```
Comment (holgerkoser) [inline frontend/src/components/ProjectUserRow.vue]: rolename as key

File: frontend/src/utils/index.js [hunk 16]
```diff
@@ -628,6 +628,22 @@ export function allErrorCodesFromLastErrors (lastErrors) {
   return uniq(compact(flatMap(lastErrors, 'codes')))
 }
 
+export const allMemberRoles = [
```
Comment (holgerkoser) [inline frontend/src/utils/index.js]: MEMBER_ROLE_DESCRIPTORS or something like this. Do we need this in utils?

File: frontend/src/store/index.js [hunk 17]
```diff
@@ -817,13 +817,20 @@ const actions = {
...
         return res
       })
   },
+  updateMember ({ dispatch, commit }, { name, roles }) {
+    return dispatch('members/update', { name, roles })
+      .then(res => {
+        dispatch('setAlert', { message: 'Member updated', type: 'success' })
+        return res
+      })
+  },
```
Comment (holgerkoser) [inline frontend/src/store/index.js]:
```suggestion
  async updateMember ({ dispatch, commit }, payload) {
    const result = await dispatch('members/update', payload)
    await dispatch('setAlert', { message: 'Member updated', type: 'success' })
    return result
  },
```

File: frontend/src/store/modules/members.js [hunk 18]
```diff
@@ -38,14 +38,20 @@ const actions = {
         throw error
       })
   },
-  add: ({ commit, rootState }, name) => {
+  add: ({ commit, rootState }, data) => {
```
Comment (holgerkoser) [inline frontend/src/store/modules/members.js]:
```suggestion
  add ({ commit, rootState }, data) {
```

File: frontend/src/store/modules/members.js [hunk 19]
```diff
@@ -38,14 +38,20 @@ const actions = {
...
     const namespace = rootState.namespace
-    const data = { name }
     return addMember({ namespace, data })
       .then(res => {
         commit('RECEIVE', res.data)
       })
   },
+  update: ({ commit, rootState }, { name, roles }) => {
```
Comment (holgerkoser) [inline frontend/src/store/modules/members.js]:
```suggestion
  update ({ commit, rootState }, { name, roles }) {
```

File: frontend/src/pages/Members.vue [hunk 20]
```diff
@@ -387,6 +397,26 @@ export default {
...
+    onEditUser (username, userroles) {
+      this.updateMemberName = username
+      this.updateMemberRoles = userroles
+      this.openUserUpdateDialog()
+    },
+    onEditServiceAccount (username, userroles) {
+      this.updateMemberName = username
+      this.updateMemberRoles = userroles
+      this.openServiceAccountUpdateDialog()
+    },
```
Comment (holgerkoser) [inline frontend/src/pages/Members.vue]:
```suggestion
    onEditUser (username, roles) {
      this.updateMemberName = username
      this.updateMemberRoles = roles
      this.openUserUpdateDialog()
    },
    onEditServiceAccount (username, roles) {
      this.updateMemberName = username
      this.updateMemberRoles = roles
      this.openServiceAccountUpdateDialog()
    },
```

File: frontend/src/pages/Members.vue [hunk 21]
```diff
@@ -387,6 +397,26 @@ export default {
...
+    roleName (role) {
+      const roleObject = find(allMemberRoles, { name: role })
+      if (roleObject) {
+        if (roleObject.hidden) {
+          return undefined
+        }
+        return roleObject.displayName
+      }
+      return upperFirst(role)
```
Comment (holgerkoser) [inline frontend/src/pages/Members.vue]:
```suggestion
    getRoleDisplayName (role) {
      const roleDescriptor = find(MEMBER_ROLE_DESCRIPTORS, ['name', role])
      if (roleDescriptor) {
        if (roleDescriptor.hidden) {
          return
        }
        return roleDescriptor.displayName
      }
      return upperFirst(role)
```

File: frontend/src/pages/Members.vue [hunk 22]
```diff
@@ -269,7 +279,7 @@ export default {
           avatarUrl: gravatarUrlGeneric(username),
           displayName: displayName(username),
           created: getTimestampFormatted(serviceAccount.creationTimestamp),
-          roleNames: map(serviceAccount.roles, this.roleName)
+          roleNames: compact(map(serviceAccount.roles, this.roleName))
```
Comment (holgerkoser) [inline frontend/src/pages/Members.vue]:
```suggestion
          roleDisplayNames: compact(map(serviceAccount.roles, this.getRoleDisplayName))
```

File: frontend/src/pages/Members.vue [hunk 23]
```diff
@@ -229,7 +239,7 @@ export default {
       serviceAccountHelpDialog: false,
       kubeconfigDialog: false,
       updateMemberName: undefined,
-      updateUserRoles: undefined,
+      updateMemberRoles: undefined,
```
Comment (holgerkoser) [inline frontend/src/pages/Members.vue]: It confuses me completely if a property has the name of a method and vice versa.  I expect `updateMemberRoles` and `updateMemberName` to be a method nd not a property. I expect `roleName` to be a property and not a method which returns the displayName for a given role.

File: frontend/src/dialogs/MemberDialog.vue [hunk 24]
```diff
@@ -88,13 +111,17 @@ export default {
       type: String,
       required: true
     },
-    oldName: {
+    currentName: {
```
Comment (holgerkoser) [inline frontend/src/dialogs/MemberDialog.vue]: use `name` and `roles` as component properties and handle the change of these values internally. The internal name could be `member.name` or `memberName` or `newName`.

File: backend/lib/services/members.js (hunk 3 above)
Comment (petersutter) [reply backend/lib/services/members.js]: gardener does not accept "empty" roles
`": ValidationError(Project.spec.members[19]): missing required field "role" in com.github.gardener.gardener.pkg.apis.core.v1beta1.ProjectMember`

File: frontend/src/dialogs/MemberDialog.vue [hunk 25]
```diff
@@ -166,42 +204,54 @@ export default {
     textField () {
       return this.$refs.name
     },
+    allMemberRoles () {
+      return compact(map(allMemberRoles, role => {
```
Comment (petersutter) [inline frontend/src/dialogs/MemberDialog.vue]: filter first

File: frontend/src/dialogs/MemberDialog.vue [hunk 26]
```diff
@@ -247,21 +310,43 @@ export default {
...
+      if (this.currentName) {
+        this.name = this.currentName
       } else if (this.isUserDialog) {
         this.name = defaultUsername
       } else if (this.isServiceDialog) {
         this.name = this.defaultServiceName()
       }
 
+      if (this.currentRoles) {
+        this.roles = this.currentRoles.slice()
```
Comment (petersutter) [inline frontend/src/dialogs/MemberDialog.vue]:
```suggestion
        this.roles = [...this.currentRoles]
```

**Comment by petersutter** [comment]
add confirmation dialog when deleting a member

File: frontend/src/store/modules/members.js [hunk 27]
```diff
@@ -38,14 +38,20 @@ const actions = {
...
+    const namespace = rootState.namespace
+    return updateMember({ namespace, name, data: { roles } })
+      .then(res => {
```
Comment (petersutter) [inline frontend/src/store/modules/members.js]: async / await

File: backend/lib/routes/members.js (hunk 1 above)
Comment (grolu) [reply backend/lib/routes/members.js]: done

File: backend/test/acceptance/api.members.spec.js (hunk 8 above)
Comment (grolu) [reply backend/test/acceptance/api.members.spec.js]: done

File: frontend/src/components/ProjectServiceAccountRow.vue (hunk 13 above)
Comment (grolu) [reply frontend/src/components/ProjectServiceAccountRow.vue]: refactored to roleDisplayName

File: frontend/src/utils/index.js (hunk 16 above)
Comment (grolu) [reply frontend/src/utils/index.js]: renamed to MEMBER_ROLE_DESCRIPTORS, keep it in utils for now will move to store when we read custom roles

File: frontend/src/pages/Members.vue (hunk 23 above)
Comment (grolu) [reply frontend/src/pages/Members.vue]: renamed to update**d**

File: frontend/src/dialogs/MemberDialog.vue (hunk 24 above)
Comment (grolu) [reply frontend/src/dialogs/MemberDialog.vue]: renamed.

**Comment by grolu** [comment]
> add confirmation dialog when deleting a member

done

File: backend/lib/utils/index.js [hunk 28]
```diff
@@ -53,10 +53,25 @@ function shootHasIssue (shoot) {
   return _.get(shoot, ['metadata', 'labels', 'shoot.garden.sapcloud.io/status'], 'healthy') !== 'healthy'
 }
 
+function toOneMemberRoleArray (role, roles) {
```
Comment (holgerkoser) [inline backend/lib/utils/index.js]:
```suggestion
function joinMemberRoleAndRoles (role, roles) {
```

File: backend/lib/utils/index.js [hunk 29]
```diff
@@ -53,10 +53,25 @@ function shootHasIssue (shoot) {
...
+  if (roles) {
+    // uniq to also support test scenarios, gardener discards duplicate roles
+    return _.uniq([role, ...roles])
+  }
+  return [role]
+}
+
+function toMemberRoleRolesArrays (roles) {
```
Comment (holgerkoser) [inline backend/lib/utils/index.js]:
```suggestion
function splitMemberRolesIntoRoleAndRoles (roles) {
```

File: frontend/src/pages/Members.vue [hunk 30]
```diff
@@ -400,17 +407,31 @@ export default {
     confirmDelete (username) {
       const memberName = displayName(username)
       const projectName = this.projectDetails.projectName
+      let messageHtml
+      if (this.isCurrentUser(username)) {
+        messageHtml = `Do you want to remove <span class="red--text text--darken-2 font-weight-bold">yourself</span> from the project <i>${projectName}</i>?`
+      } else {
+        messageHtml = `Do you want to delete the member <i>${memberName}</i> from the project <i>${projectName}</i>?`
+      }
```
Comment (petersutter) [inline frontend/src/pages/Members.vue]: we should escape the variable input (we missed it initially)

**Comment by grolu** [comment]
> apart from the names `toOneMemberRoleArray` and `toMemberRoleRolesArrays` this PR lgtm. At least `...RoleRolesArrays` makes no sense.

Yes maybe I was a bit too creative ;)
